### PR TITLE
Image Clustering Power Skill using DBSCAN

### PR DESCRIPTION
**Pull request by shanepeckham**
## Purpose

* ...
This Power Skill will uses the [DBSCAN](https://scikit-learn.org/stable/modules/generated/sklearn.cluster.dbscan.html)
unsupervised clustering algorithm alongside [VGG16](https://keras.io/api/applications/vgg/) to extract
visual features and cluster images. 

This skill is ideal for:

1) Exploring your data to identify clusters based on visual features during your data exploration phase
2) Using in conjunction with Custom Vision Classification to further cluster your images, for example if
you need a hierarchical classification structure.
3) Auto-labelling your images based on the clusters identified and the labels you associated with the clusters. Note
Azure Machine Learning has an auto-labelling feature already, this Power Skill should be used if this feature is not suitable
## Does this introduce a breaking change?

```
[ ] Yes
[ X] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[ X] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## How to Test
*  Get the code

```
git clone [repo-address]
cd [repo-name]
git checkout [branch-name]
npm install
```

* Test the code

```
The first step in the process is to extract VGG16 embeddings from the images and train the DBSCAN model on the extracted features.
To better understand the algorithm itself, please use explanatory notebook, it contains a local example of the process.
To leverage the power of Azure Machine Learning, train a model on a cluster and save trained model into registry to make it available for your custom skill, you can either use training notebook or proceed with setting up the full MLOps process using clustering pipeline.
As with any (especially, unsupervised) machine learning solution, inspecting the clusters generated and playing with the algorithm hyperparameters will be required.
To explore generated clusters and generate labels dictionary required for the custom skill, you can use labeling notebook. These labels are what will be indexed to retrieve the images.
Clusters report is also available under the registered model on the Azure Machine Learning Portal.
Deploy the skill and add the endpoint to your skillset file
Run your indexer [deployment/azuresearch/create_indexer.json]
Investigate your indexed data and compare the effect of using Image Clustering Power Skill and Computer Vision Service using Azure Search notebook.
```

## What to Check
Verify that the following are valid
* ...

### CODE REVIEW

File: Vision/ImageClusteringSkill/README.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
+# Image Clustering Power Skill #
+
+This Power Skill will uses the [DBSCAN](https://scikit-learn.org/stable/modules/generated/sklearn.cluster.dbscan.html)
```
Comment (vkurpad) [inline Vision/ImageClusteringSkill/README.md:3]: Remove will

File: Vision/ImageClusteringSkill/README.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
+# Image Clustering Power Skill #
+
+This Power Skill will uses the [DBSCAN](https://scikit-learn.org/stable/modules/generated/sklearn.cluster.dbscan.html)
+unsupervised clustering algorithm alongside [VGG16](https://keras.io/api/applications/vgg/) to extract
+visual features and cluster images. 
+
+This skill is ideal for:
```
Comment (vkurpad) [inline Vision/ImageClusteringSkill/README.md:7]: Could you add one visual example? A sample image that gets tagged

File: Vision/ImageClusteringSkill/README.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+   * To leverage the power of [Azure Machine Learning](https://azure.microsoft.com/en-us/services/machine-learning/), train a model on a cluster and save trained model into registry to make it available for your custom skill, you can either use [training notebook](notebooks/2-aml-training-pipeline.ipynb) or proceed with setting up the full [MLOps process](https://github.com/microsoft/MLOpsPython/tree/master) using [clustering pipeline](mlops/clustering_pipeline).
+2) As with any (especially, unsupervised) machine learning solution, inspecting the clusters generated and playing with the algorithm hyperparameters will be required.
+   * To explore generated clusters and generate labels dictionary required for the custom skill, you can use [labeling notebook](notebooks/3-label-clusters.ipynb). These labels are what will be indexed to retrieve the images.
+   * Clusters report is also available under the registered model on the Azure Machine Learning Portal.  
+3) Deploy the skill and add the endpoint to your [skillset file](deployment/azuresearch/create_skillset.json)
```
Comment (vkurpad) [inline Vision/ImageClusteringSkill/README.md:33]: No immediate action needed, but it would be good to convert the is step to use the AML skill in a future PR.

**Review by vkurpad — COMMENTED** [review]
I am in the process of deploying and testing the skill, will post any feedback if I run into issues.

**Review by bleroy — CHANGES_REQUESTED** [review]
Thanks for the contribution. My general concern is that the signal to noise ratio is very small. There's a lot of boilerplate, plumbing, and sample files. It's hard for a newcomer to figure out where the relevant parts are and how to modify them to their needs. The weight of this change is also several times all the other skills in the repository combined.
I think in order to be more useful, it would be better to separate the actual skills and move the rest to optional zip files, documentation explaining how to build an deploy, etc.

**Comment by bleroy** [comment]
There's also a few empty readme files that should be removed, and .DS_Store files should be excluded.

File: Vision/ImageClusteringSkill/ml/models/search_service.py [hunk 4]
```diff
@@ -0,0 +1,583 @@
+import requests
+import json
+from typing import List, Dict
+
+class SearchService:
```
[the remark this answers is not in the document]
Comment (shanepeckham) [reply Vision/ImageClusteringSkill/ml/models/search_service.py]: This has now been removed

File: Template/PythonFastAPI/README.md [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+- Pip
+
+Once your container is up and running you should:
+
+1. Open your test `.py` file (```tests/powerskill_api_test.py```) and set the Python interpreter to be your venv (bottom blue bar of VSCode)
+2. Use the python test explorer plugin to run your tests or click the 'run test' prompt above your tests
+
+## Deploy to Azure
+
+The infra folder contains terraform files to deploy this to FastAPI container to Web
```
[the remark this answers is not in the document]
Comment (shanepeckham) [reply Template/PythonFastAPI/README.md:84]: Added
